File: infixToPostfix.java
import java.util.*;
import java.lang.*;
import java.io.*;

class infixToPostfix
{
    static Scanner scanner = new Scanner(System.in);
    public static void main (String[] args)
	{
	    int T = Integer.parseInt(scanner.nextLine());
	    for(int a=0; a<T; a++)
	    {
	        String input = scanner.nextLine();
	        
	        doInfixToPostfix(input);
	    }
	}
	
	static void doInfixToPostfix(String input)
	{
	    Stack<Character> stack = new Stack<>();
	    
	    String result = "";
	    for(int i=0; i<input.length(); i++)
	    {
	        char c = input.charAt(i);
	        int cInt = (int) c;
	        
	        if((cInt>=65 && c<=90) || (cInt>=97 && c<=122) || (cInt>=48 && c<=57))
	        {
	            result += c;
	        }
	        else
	        {
	            if(stack.empty())
	            {
	                stack.push(c);
	            }
	            else if(c == '(') //opening bracket
				{
					stack.push(c);
				}
				else if(c == ')') //closing bracket
				{
					char top = stack.peek();
					
					while(!stack.empty() && top != '(')
					{
					    result += top;
					    stack.pop();
					    
					    if(!stack.empty())
						{
							top = stack.peek();
						}
					}
					stack.pop(); // lastly to remove open bracket
				}
				else
				{
				    char top = stack.peek();
				    
				    while(!stack.empty() && isLower(c, top))
				    {
				        result += top;
					    stack.pop();
					    
					    if(!stack.empty())
						{
							top = stack.peek();
						}
				    }
				    stack.push(c); //lastly pushing that c
				}
	        }
	    }
	    
    //if some operators are left in stack then popping them	
		while(!stack.empty()) //t is lower than top
		{
			char top = stack.pop();
			result += top;
		}
	    
	    System.out.println(result);
	}
	
	public static int weight(char c) // return the weight of the opeartor
	{
		int weight = -1;
		if( c =='-')
		{
			weight = 1;
		}
		else if( c =='+')
		{
			weight = 1;
		}
		else if( c =='*')
		{
			weight = 2;
		}
		else if( c =='/')
		{
			weight = 3;
		}
		else if( c =='^')
		{
			weight = 4;
		}

		return weight;
	}

	public static boolean isLower(char fi, char sec) // return true if fi has lower presedence than sec
	{
		int a = weight(fi);
		int b = weight(sec);
			
		return (a<=b);
	}

}

File: JS/infixToPostfix.js
Array.prototype.isEmpty = function() {
    return this.length == 0;
}

Array.prototype.peek = function() {
    return this[this.length - 1];
}

function isOperator(char) {
    return ['+', '-', '*', '/', '^', '(', ')'].includes(char)
}

function getWeight(char) {
    if (["+", "-"].includes(char)) {
        return 1;
    } else if (["*"].includes(char)) {
        return 2;
    } else if (["/"].includes(char)) {
        return 3;
    } else if (["^"].includes(char)) {
        return 4;
    }

    return 0;
}

function doCalculation(no1, no2, operator) {
    if (operator == "+") {
        return no1 + no2
    } else if (operator == "-") {
        return no1 - no2
    } else if (operator == "*") {
        return no1 * no2
    } else if (operator == "/") {
        return no1 / no2
    } else if (operator == "^") {
        return no1 ^ no2
    }

    return 0;
}

function infixToPostfix(str) {
    let stack = [];

    let postfix = "";
    for (let i = 0; i < str.length; i++) {
        let char = str[i].trim();
        if (!char) continue; //if character is space

        if (!isOperator(char)) {
            postfix += char;
        } else {
            if (stack.isEmpty()) {
                stack.push(char);
            } else if (char == '(') {
                // opening bracket
                stack.push(char);
            } else if (char == ')') {
                // closing bracket
                
                let top = stack.peek();
                while ((top != "(") && !stack.isEmpty()) {
                    postfix += top;
                    stack.pop();

                    if (!stack.isEmpty()) top = stack.peek();
                }
                stack.pop(); // lastly to remove open bracket `(`
            } else {
                // if top of the stack has less weight than the operator then push the operator in the stack
                // otherwise keep popping till in stack any element with greater weight appears

                let top = stack.peek();
                while ((getWeight(char) <= getWeight(top)) && !stack.isEmpty()) {
                    postfix += top;
                    stack.pop();

                    if (!stack.isEmpty()) top = stack.peek();
                }

                stack.push(char);
            }
        }
    }

    while (!stack.isEmpty()) {
        let topPop = stack.pop();
        postfix += topPop;
    }

    return postfix;
}

let infix = "a+b*(c^d-e)^(f+g*h)-i"; // "A+B*C-D*E";
console.log("infix", infix)

let postfix = infixToPostfix(infix);
console.log("postfix", postfix)